python/uv/private/lock.bzl

```diff
@@ -40,86 +276,78 @@ def lock(*, name, srcs, out, upgrade = False, universal = True, args = [], **kwa
+    maybe_out = _maybe_file(out)
+    out_new = out + ".new"
+    target_compatible_with = kwargs.pop("target_compatible_with", select({
+        "@platforms//os:windows": ["@platforms//:incompatible"],
+        "//conditions:default": [],
+    })) if BZLMOD_ENABLED else ["@platforms//:incompatible"]
+
+    _lock(
         name = name,
         srcs = srcs,
-        outs = [out + ".new"],
-        cmd_bash = cmd,
+        # Check if the output file already exists, if yes, first copy it to the
+        # output file location in order to make `uv` not change the requirements if
+        # we are just running the command.
+        existing_output = maybe_out,
+        update_target = update_target,
+        output = out_new,
         tags = [
             "local",
             "manual",
             "no-cache",
+            "requires-network",
         ],
-        target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
-        toolchains = [
-            Label("//python/uv:current_toolchain"),
-            Label("//python:current_py_toolchain"),
-        ],
+        args = args,
+        env = env,
+        target_compatible_with = target_compatible_with,
+    )
+
+    # A target for updating the in-tree version directly by skipping the in-action
+    # uv pip compile.
+    _lock_run(
+        name = locker_target,
+        lock = name,
+        # TODO @aignas 2025-03-13: should we actually update the in-tree
+        # version with this? For now I think that yes because we may want to
+        # specify extra args like `--upgrade` or `-P mypkg` to upgrade some of
+        # the packages.
+        output = out,
+        # TODO @aignas 2025-03-13: allow customizing the env
+        # env = env,
+        tags = ["manual"],
     )
 
     # Write a script that can be used for updating the in-tree version of the
     # requirements file
-    write_file(
-        name = name + ".update_gen",
+    pkg = native.package_name()
+    expand_template(
+        name = update_target + "_gen",
         out = update_target + ".py",
-        content = [
-            "from os import environ",
-            "from pathlib import Path",
-            "from sys import stderr",
-            "",
-            'src = Path(environ["REQUIREMENTS_FILE"])',
-            'assert src.exists(), f"the {src} file does not exist"',
-            'dst = Path(environ["BUILD_WORKSPACE_DIRECTORY"]) / "{}" / "{}"'.format(pkg, out),
-            'print(f"Writing requirements contents\\n  from {src.absolute()}\\n  to {dst.absolute()}", file=stderr)',
-            "dst.write_text(src.read_text())",
-            'print("Success!", file=stderr)',
-        ],
+        template = "//python/uv/private:copy.py",
+        substitutions = {
+            'dst = ""': 'dst = "{}/{}"'.format(pkg, out),
+            'src = ""': 'src = "{}/{}"'.format(pkg, out_new),
+        },
     )
 
-    py_binary(
+    # TODO @aignas 2025-03-13: create a py_test-like target if the design of
```

Comment:
TBH, I don't understand the point of the diff test. Dependencies change pretty regularly it just seems like a source of spantaneous churn?

A test that verifies requirements.txt came from the right requirements.in version sounds appealing (reminds you to re-run locking), but, doesn't the diff test perform locking again?

Reply:
Yeah, the test is something that exists today and I am not sure about the usefulness of it either. I'll include it when I can address the changes, but we can discuss later if it is really needed. It can be opt-in target, where users can use native_test or something similar if they want.